src/program/game/systems/system/system.abstract.ts

```diff
@@ -17,76 +16,118 @@ export abstract class SystemAbstract {
         this.lastUpdateEntityIdList = [];
     }
 
-    protected abstract initEntity(entity: EntityAbstract);
-    protected abstract updateEntity(delta: number, entity: EntityAbstract);
-    protected abstract onDataEntityUpdate(
+    /**
+     * Called when an entity is available on the system
+     * @param entity
+     * @protected
+     */
+    protected onInitEntity(entity: EntityAbstract) {};
+
+    /**
+     * Called every frame
+     * @param delta
+     * @param entity
+     * @protected
+     */
+    protected onUpdateEntity(delta: number, entity: EntityAbstract) {};
+
+    /**
+     * Called when some of the entity data is changed
+     * @param entity
+     * @param componentEnums: Array {ComponentEnum[]} containing every componentEnum changed
+     * @param oldEntityData
+     * @param newEntityData
+     * @protected
+     */
+    protected onDataEntityUpdate(
         entity: EntityAbstract,
         componentEnums: ComponentEnum[],
         oldEntityData: ComponentTypes,
         newEntityData: ComponentTypes
-    );
-    protected abstract deleteEntity(entity: EntityAbstract);
+    ) {};
+
+    /**
+     * Called when the entity is removed from the system
+     * @param entity
+     * @protected
+     */
+    protected onRemoveEntity(entity: EntityAbstract) {};
+
+    /**
+     * Returns {string[]} containing all the entities id
+     * @protected
+     */
+    protected getEntityIdList(): string[] {
+        const { components } = Program.getInstance().store.getState();
+        return this.components.map(componentEnum => components[componentEnum].entities).flat(2);
+    }
+
+    /**
+     * Returns {EntityAbstract} or undefined if the entity doesn't exists on the current system
+     * @param id
+     * @public
+     */
+    public getEntity(id: string): EntityAbstract | undefined {
+        return this.getEntities().find(entity => entity?.id === id);
+    }
 
-    public update(delta: number) {
+    /**
+     * Returns {EntityAbstract[]} with all the entities inside the system.
+     * @public
+     */
+    public getEntities(): EntityAbstract[] {
+        const program = Program.getInstance();
+        const entityList = this.getEntityIdList();
+        const entityCountMap = new Map([...new Set(entityList)].map(x => [x, entityList.filter(y => y === x).length] ));
+        return [...new Set(entityList.filter(o => entityCountMap.get(o) === this.components.length))]
+            .map(entityId => program.game.entities.get(entityId))
+            .filter(entity => entity);
+    }
+
+    public _update(delta: number) {
```

Comment:
Why _update? If it is a private property we can use the `private` keyword instead

Reply:
This update function is called outside, but it's not pretended to be called inside the extended class.

Reply:
ah ok, I understand now, notation `_method` was a bit weird to me. But it is alright now